<p style="text-align:right;">
<img src="https://upload.wikimedia.org/wikipedia/commons/c/c2/WHO_logo.svg"
     alt="DigitalFuturesLogo"
     style="float: center; margin-right: 10px;"
     width="100"
     />
</p>

**<h1 style="text-align: center;">World Health Organisation</h1>**
**<h2 style="text-align: center;">Life Expectancy Measurements</h2>**
**<h3 style="text-align: center;">Regression Rangers</h3>**
**<h4 style="text-align: center;">Riya, Dewi, Zach, Kal</h4>**
**<h5 style="text-align: left;">15-02-2024</h5>**

### **Importing Libraries**

In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split

import statsmodels.api as sm
import statsmodels.tools

from statsmodels.stats.outliers_influence import variance_inflation_factor

import seaborn as sns
import matplotlib.pyplot as plt

### **1. Import the data**

In [2]:
# Importing the data file.

path = "Life Expectancy Data.csv"
whodf = pd.read_csv(path)

### **2. Feature Engineering**

In [3]:
# Creating a list of feature columns and remove the unnecessary columns

feature_cols = list(whodf.columns)
feature_cols.remove('Country') # Not Enough Data Per Entry
feature_cols.remove('Life_expectancy') # Target
feature_cols.remove('Economy_status_Developed') # Inverse of 'Economy_status_Developing'

In [4]:
# Create X for feature, and y for target
# train_test_split

X = whodf[feature_cols]
y = whodf['Life_expectancy']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [5]:
# Create the function for feature engineering

def feature_eng(df):
        df = df.copy()
        df = pd.get_dummies(df, columns = ['Region'], drop_first = True, prefix = 'region', dtype=int)
        df['GDP_per_capita_log'] =  np.log(df['GDP_per_capita'])
        df = sm.add_constant(df)
        return df

In [6]:
## Transform using FE func

X_train_fe = feature_eng(X_train)
X_test_fe = feature_eng(X_test)

In [7]:
feature_cols = list(X_train_fe.columns)

### **3. Linear Regresion**

We start our investigation by looking at a linear regression including all of the feature variables:

In [8]:
## Create a linear regression model

fullresult = feature_cols
lin_reg = sm.OLS(y_train, X_train_fe[fullresult])
fullresults = lin_reg.fit()
fullresults.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_expectancy   R-squared:                       0.985
Model:                            OLS   Adj. R-squared:                  0.984
Method:                 Least Squares   F-statistic:                     5554.
Date:                Thu, 15 Feb 2024   Prob (F-statistic):               0.00
Time:                        12:18:57   Log-Likelihood:                -3624.5
No. Observations:                2291   AIC:                             7303.
Df Residuals:                    2264   BIC:                             7458.
Df Model:                          26                                         
Covariance Type:            nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
const                                    4.2991     11.613      0.370      0.711     -18.474      27.072
Year                                     0.0394      0.006      6.835      0.000       0.028       0.051
Infant_deaths                           -0.0516      0.006     -8.108      0.000      -0.064      -0.039
Under_five_deaths                       -0.0463      0.004    -11.677      0.000      -0.054      -0.039
Adult_mortality                         -0.0459      0.001    -72.912      0.000      -0.047      -0.045
Alcohol_consumption                     -0.0225      0.012     -1.915      0.056      -0.045       0.001
Hepatitis_B                             -0.0076      0.003     -3.002      0.003      -0.013      -0.003
Measles                                  0.0011      0.002      0.620      0.535      -0.002       0.005
BMI                                     -0.1560      0.023     -6.748      0.000      -0.201      -0.111
Polio                                    0.0136      0.006      2.319      0.020       0.002       0.025
Diphtheria                              -0.0074      0.006     -1.268      0.205      -0.019       0.004
Incidents_HIV                            0.0743      0.019      3.978      0.000       0.038       0.111
GDP_per_capita                        5.989e-06   3.27e-06      1.830      0.067    -4.3e-07    1.24e-05
Population_mln                          -0.0002      0.000     -1.167      0.243      -0.001       0.000
Thinness_ten_nineteen_years             -0.0363      0.017     -2.187      0.029      -0.069      -0.004
Thinness_five_nine_years                 0.0221      0.017      1.338      0.181      -0.010       0.055
Schooling                                0.0869      0.019      4.692      0.000       0.051       0.123
Economy_status_Developing               -2.5244      0.165    -15.333      0.000      -2.847      -2.202
region_Asia                              0.4683      0.113      4.135      0.000       0.246       0.690
region_Central America and Caribbean     2.0946      0.122     17.163      0.000       1.855       2.334
region_European Union                   -0.4597      0.178     -2.584      0.010      -0.809      -0.111
region_Middle East                       0.2609      0.139      1.870      0.062      -0.013       0.534
region_North America                     0.8872      0.234      3.793      0.000       0.428       1.346
region_Oceania                          -0.6099      0.146     -4.171      0.000      -0.897      -0.323
region_Rest of Europe                    0.6003      0.145      4.133      0.000       0.315       0.885
region_South America                     1.8163      0.135     13.461      0.000       1.552       2.081
GDP_per_capita_log                    

In [9]:
print(fullresult)

['const', 'Year', 'Infant_deaths', 'Under_five_deaths', 'Adult_mortality', 'Alcohol_consumption', 'Hepatitis_B', 'Measles', 'BMI', 'Polio', 'Diphtheria', 'Incidents_HIV', 'GDP_per_capita', 'Population_mln', 'Thinness_ten_nineteen_years', 'Thinness_five_nine_years', 'Schooling', 'Economy_status_Developing', 'region_Asia', 'region_Central America and Caribbean', 'region_European Union', 'region_Middle East', 'region_North America', 'region_Oceania', 'region_Rest of Europe', 'region_South America', 'GDP_per_capita_log']


In [10]:
# Calculating the RMSE

y_pred = fullresults.predict(X_train_fe[fullresult])

rmse = statsmodels.tools.eval_measures.rmse(y_train, y_pred)

print(rmse)

1.1771727873191782


In [11]:
# Testing on test data and calculating the RMSE

y_test_pred = fullresults.predict(X_test_fe[fullresult])

rmse = statsmodels.tools.eval_measures.rmse(y_test, y_test_pred)

print(rmse)

1.212457889546913


This model produces very accurate predictions, with an average error of only 1.2 years. However, this regression raises a warning on the condition number, implying that there are multicollinearity problems. This leads to a less robust and reliable prediction, and we will need to select some variables to reduce this. 

### **4. P-value Optimisation & Stepwise Condition Number Reduction**

In order to deal with this, we are going to select the most robust predictors of life expectancy by selecting the ones with the smallest p-value, but only add them in if they do not increase the condition number by too much. In this way, we balance our prediction accuracy with the robustness of the model.

**First the parameters are ordered by their P-value \
Then they are split into a base list of required variables and the potential variables**

In [12]:
full_ordered = pd.DataFrame(fullresults.pvalues).to_dict()[0]
full_ordered = sorted(full_ordered, key = full_ordered.get)

base = ['const', 'Adult_mortality', 'Infant_deaths', 'Under_five_deaths']
ordered = []
for i in full_ordered:
    if i not in base:
        ordered.append(i)
ordered

['region_Central America and Caribbean',
 'Economy_status_Developing',
 'region_South America',
 'Year',
 'BMI',
 'GDP_per_capita_log',
 'Schooling',
 'region_Oceania',
 'region_Asia',
 'region_Rest of Europe',
 'Incidents_HIV',
 'region_North America',
 'Hepatitis_B',
 'region_European Union',
 'Polio',
 'Thinness_ten_nineteen_years',
 'Alcohol_consumption',
 'region_Middle East',
 'GDP_per_capita',
 'Thinness_five_nine_years',
 'Diphtheria',
 'Population_mln',
 'Measles']

In [13]:
def stepwise_cond_reduction(base_features, ordered_features, y, X, threshold = 100, upper_limit = 1000):
    # Has a base set of features to use
    new_features = base_features
    
    # Calculates the base Condition Number
    lin_reg = sm.OLS(y, X[new_features])
    results = lin_reg.fit()
    holder = results.condition_number
    
    # For each feature in order it checks if adding it keeps it below a wanted threshold
    ## If so adds, if not removes
    for i in ordered_features:
        temp_features = new_features.copy()
        temp_features.append(i)
        lin_reg = sm.OLS(y, X[temp_features])
        results = lin_reg.fit()
        if results.condition_number < holder + threshold:
            if results.condition_number < upper_limit:
                new_features = temp_features
                holder = results.condition_number
    return new_features

**The base variables can be changed but should contain the constant. \
The upper limit ensures there is no warning from too high a Condition Number \
Increasing the threshold can allow more variables with a ow P-value which increase the Condition Number more and vice versa**

In [14]:
main_result = stepwise_cond_reduction(
    base,
    full_ordered,
    y_train, X_train_fe, 300)
main_result

['const',
 'Adult_mortality',
 'Infant_deaths',
 'Under_five_deaths',
 'region_Central America and Caribbean',
 'Economy_status_Developing',
 'region_Asia',
 'Incidents_HIV',
 'Thinness_ten_nineteen_years',
 'Thinness_five_nine_years',
 'Population_mln']

**This model can be truncated if the parameters towards the end do not imporve the model significantly**

In [15]:
# Removing the features

main_result.remove('Population_mln')
main_result.remove('Thinness_five_nine_years')

In [16]:
## Create a linear regression model

lin_reg = sm.OLS(y_train, X_train_fe[main_result])
main_results = lin_reg.fit()
main_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_expectancy   R-squared:                       0.980
Model:                            OLS   Adj. R-squared:                  0.980
Method:                 Least Squares   F-statistic:                 1.370e+04
Date:                Thu, 15 Feb 2024   Prob (F-statistic):               0.00
Time:                        12:18:57   Log-Likelihood:                -3943.9
No. Observations:                2291   AIC:                             7906.
Df Residuals:                    2282   BIC:                             7957.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
const                                   83.7301      0.082   1021.789      0.000      83.569      83.891
Adult_mortality                         -0.0481      0.001    -77.216      0.000      -0.049      -0.047
Infant_deaths                           -0.0741      0.007    -10.742      0.000      -0.088      -0.061
Under_five_deaths                       -0.0399      0.004     -9.447      0.000      -0.048      -0.032
region_Central America and Caribbean     1.5727      0.101     15.506      0.000       1.374       1.772
Economy_status_Developing               -2.3258      0.091    -25.669      0.000      -2.503      -2.148
region_Asia                              0.4320      0.095      4.561      0.000       0.246       0.618
Incidents_HIV                            0.1200      0.019      6.263      0.000       0.082       0.158
Thinness_ten_nineteen_years             -0.0310      0.008     -3.703      0.000      -0.047      -0.015
==============================================================================
Omnibus:                        6.036   Durbin-Watson:                   2.002
Prob(Omnibus):                  0.049   Jarque-Bera (JB):                6.519
Skew:                          -0.072   Prob(JB):                       0.0384
Kurtosis:                       3.218   Cond. No.                         947.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [17]:
print(main_result)

['const', 'Adult_mortality', 'Infant_deaths', 'Under_five_deaths', 'region_Central America and Caribbean', 'Economy_status_Developing', 'region_Asia', 'Incidents_HIV', 'Thinness_ten_nineteen_years']


In [18]:
# Testing on train data and calculating the RMSE

y_pred = main_results.predict(X_train_fe[main_result])

rmse = statsmodels.tools.eval_measures.rmse(y_train, y_pred)

print(rmse)

1.353259817356647


In [19]:
# Testing on test data and calculating the RMSE

y_test_pred = main_results.predict(X_test_fe[main_result])

rmse = statsmodels.tools.eval_measures.rmse(y_test, y_test_pred)

print(rmse)

1.384123274348449


What we have produced is a model that has slightly worse prediction accuracy (1.38 years vs 1.22 years), but no longer raises a warning over the multicollinearity of our features. We think that this is a good balance between accuracy and reliability of the model.

### **5. Limited Parameter Model**

The client has raised the potential issue surrounding the inclusion of medical data, as this could bring about unwanted financial implications. For this reason,  we chose to omit data that could be related to healthcare, healthcare financing and infrastructure such as vaccination and disease data. Furthermore, we dropped all data relating to child mortality, as this necessarily makes implications on the state of government infrastructure. 

On the other hand, we include the Adult mortality rate as it is a lot harder to draw reasonable conclusions from that data without a proper investigation. This is because the data is uncategorized and includes Adults under 60 which is a very general scope. Furthermore, without the specifics, we can not directly relate it to either healthcare or quality of life statistics. We also include general geographic and macroeconomic data, which does not raise privacy concerns as it is public information.

Adult Mortality and geographic variables have proved to be important throughout our investigation, hence in a limited model we think they are worthwhile inclusions.

In [20]:
chosen_result = ['const','Adult_mortality',
 'region_Central America and Caribbean', 'region_Rest of Europe', 'region_Oceania',
 'Economy_status_Developing']
lin_reg = sm.OLS(y_train, X_train_fe[chosen_result])
chosen_results = lin_reg.fit()
chosen_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_expectancy   R-squared:                       0.923
Model:                            OLS   Adj. R-squared:                  0.923
Method:                 Least Squares   F-statistic:                     5516.
Date:                Thu, 15 Feb 2024   Prob (F-statistic):               0.00
Time:                        12:18:57   Log-Likelihood:                -5458.2
No. Observations:                2291   AIC:                         1.093e+04
Df Residuals:                    2285   BIC:                         1.096e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
const                                   85.0493      0.136    625.976      0.000      84.783      85.316
Adult_mortality                         -0.0702      0.001   -128.612      0.000      -0.071      -0.069
region_Central America and Caribbean     2.8692      0.192     14.933      0.000       2.492       3.246
region_Rest of Europe                    2.1190      0.199     10.630      0.000       1.728       2.510
region_Oceania                          -0.8800      0.230     -3.827      0.000      -1.331      -0.429
Economy_status_Developing               -3.9282      0.155    -25.307      0.000      -4.233      -3.624
==============================================================================
Omnibus:                      108.305   Durbin-Watson:                   1.983
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              332.595
Skew:                          -0.151   Prob(JB):                     6.00e-73
Kurtosis:                       4.842   Cond. No.                         993.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [21]:
print(chosen_result)

['const', 'Adult_mortality', 'region_Central America and Caribbean', 'region_Rest of Europe', 'region_Oceania', 'Economy_status_Developing']


In [22]:
# Testing on train data and calculating the RMSE

y_pred = chosen_results.predict(X_train_fe[chosen_result])

rmse = statsmodels.tools.eval_measures.rmse(y_train, y_pred)

print(rmse)

2.6209211126844454


In [23]:
# Testing on test data and calculating the RMSE

y_test_pred = chosen_results.predict(X_test_fe[chosen_result])

rmse = statsmodels.tools.eval_measures.rmse(y_test, y_test_pred)

print(rmse)

2.6399724687989057


This model is less accurate than the full model (2.64 years average error). However, it still beats the benchmark of 4.5 years error, and meets the ethical requirements set out in the brief.